arangod/RocksDBEngine/RocksDBIterators.cpp

```diff
@@ -209,7 +209,25 @@ class RocksDBAllIndexIterator final : public IndexIterator {
             TRI_ASSERT(ro.snapshot != nullptr);
             TRI_ASSERT(ro.prefix_same_as_start);
             ro.verify_checksums = false;  // TODO evaluate
-            ro.iterate_upper_bound = &_upperBound;
+            // note: iterate_lower_bound/iterate_upper_bound should only be
+            // set if the iterator is not supposed to check the bounds
+            // for every operation.
+            // when the iterator is a db snapshot-based iterator, it is ok
+            // to set iterate_lower_bound/iterate_upper_bound, because this
+            // is well supported by RocksDB.
+            // if the iterator is a multi-level iterator that merges data from
+            // the db snapshot with data from an ongoing in-memory transaction
+            // (contained in a WriteBatchWithIndex, WBWI), then RocksDB does
+            // not properly support the bounds checking using
+            // iterate_lower_bound/ iterate_upper_bound. in this case we must
+            // avoid setting the bounds here and rely on our own bounds checking
+            // using the comparator. at least one underlying issue was fixed in
```

Comment:
Why cannot we always rely on "our own bounds checking using the comparator"?

Reply:
Because it is much more expensive.

Reply:
In case `iterate_upper_bound` / `iterate_lower_bound` are set, then RocksDB can filter out things on the .sst file level already, and will do some comparisons of keys against `iterate_upper_bound` / `iterate_lower_bound` itself using the comparator.

So setting the bounds is normally a good idea.
Unfortunately it is not supported for iterators that refer to the WriteBatchWithIndex (WBWI). This was a bug that we reported to RocksDB years ago, and it was only recently fixed for RocksDB 8.8, which were are not yet using.